File: bindings/rust/src/virt.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//

//! Virtualization types for the proven-servers ABI.
//!
//! Formally verified virtualization/hypervisor types.
//! Mirrors the Idris2 module `VirtABI.Types`.
//!
//! - `VmState` -- VM lifecycle states.
//! - `VirtOperation` -- VM operations.
//! - `DiskFormat` -- Virtual disk formats.
//! - `NetworkType` -- VM network types.
//! - `BootDevice` -- VM boot devices.
//!
//! All discriminant values match the Idris2 ABI tag definitions exactly.

use std::fmt;

// ===========================================================================
// VmState (tags 0-7)
// ===========================================================================

/// VM lifecycle states.
///
/// Matches `VmState` in `VirtABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum VmState {
    /// Creating (tag 0).
    Creating = 0,
    /// Running (tag 1).
    Running = 1,
    /// Paused (tag 2).
    Paused = 2,
    /// Suspended (tag 3).
    Suspended = 3,
    /// ShuttingDown (tag 4).
    ShuttingDown = 4,
    /// Stopped (tag 5).
    Stopped = 5,
    /// Crashed (tag 6).
    Crashed = 6,
    /// Migrating (tag 7).
    Migrating = 7,
}

impl VmState {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Creating),
            1 => Some(Self::Running),
            2 => Some(Self::Paused),
            3 => Some(Self::Suspended),
            4 => Some(Self::ShuttingDown),
            5 => Some(Self::Stopped),
            6 => Some(Self::Crashed),
            7 => Some(Self::Migrating),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [VmState; 8] = [
        Self::Creating, Self::Running, Self::Paused, Self::Suspended, Self::ShuttingDown, Self::Stopped, Self::Crashed, Self::Migrating,
    ];
}

impl fmt::Display for VmState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// VirtOperation (tags 0-10)
// ===========================================================================

/// VM operations.
///
/// Matches `VirtOperation` in `VirtABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum VirtOperation {
    /// Create (tag 0).
    Create = 0,
    /// Start (tag 1).
    Start = 1,
    /// Stop (tag 2).
    Stop = 2,
    /// Restart (tag 3).
    Restart = 3,
    /// Pause (tag 4).
    Pause = 4,
    /// Resume (tag 5).
    Resume = 5,
    /// Suspend (tag 6).
    Suspend = 6,
    /// Migrate (tag 7).
    Migrate = 7,
    /// Snapshot (tag 8).
    Snapshot = 8,
    /// Clone (tag 9).
    Clone = 9,
    /// Delete (tag 10).
    Delete = 10,
}

impl VirtOperation {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Create),
            1 => Some(Self::Start),
            2 => Some(Self::Stop),
            3 => Some(Self::Restart),
            4 => Some(Self::Pause),
            5 => Some(Self::Resume),
            6 => Some(Self::Suspend),
            7 => Some(Self::Migrate),
            8 => Some(Self::Snapshot),
            9 => Some(Self::Clone),
            10 => Some(Self::Delete),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [VirtOperation; 11] = [
        Self::Create, Self::Start, Self::Stop, Self::Restart, Self::Pause, Self::Resume, Self::Suspend, Self::Migrate, Self::Snapshot, Self::Clone, Self::Delete,
    ];
}

impl fmt::Display for VirtOperation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// DiskFormat (tags 0-4)
// ===========================================================================

/// Virtual disk formats.
///
/// Matches `DiskFormat` in `VirtABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum DiskFormat {
    /// Raw (tag 0).
    Raw = 0,
    /// QCOW2 (tag 1).
    Qcow2 = 1,
    /// VDI (tag 2).
    Vdi = 2,
    /// VMDK (tag 3).
    Vmdk = 3,
    /// VHD (tag 4).
    Vhd = 4,
}

impl DiskFormat {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Raw),
            1 => Some(Self::Qcow2),
            2 => Some(Self::Vdi),
            3 => Some(Self::Vmdk),
            4 => Some(Self::Vhd),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [DiskFormat; 5] = [
        Self::Raw, Self::Qcow2, Self::Vdi, Self::Vmdk, Self::Vhd,
    ];
}

impl fmt::Display for DiskFormat {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// NetworkType (tags 0-3)
// ===========================================================================

/// VM network types.
///
/// Matches `NetworkType` in `VirtABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum NetworkType {
    /// NAT (tag 0).
    Nat = 0,
    /// Bridged (tag 1).
    Bridged = 1,
    /// Internal (tag 2).
    Internal = 2,
    /// HostOnly (tag 3).
    HostOnly = 3,
}

impl NetworkType {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Nat),
            1 => Some(Self::Bridged),
            2 => Some(Self::Internal),
            3 => Some(Self::HostOnly),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [NetworkType; 4] = [
        Self::Nat, Self::Bridged, Self::Internal, Self::HostOnly,
    ];
}

impl fmt::Display for NetworkType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// BootDevice (tags 0-3)
// ===========================================================================

/// VM boot devices.
///
/// Matches `BootDevice` in `VirtABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum BootDevice {
    /// HardDisk (tag 0).
    HardDisk = 0,
    /// CD-ROM (tag 1).
    Cdrom = 1,
    /// Network (tag 2).
    Network = 2,
    /// USB (tag 3).
    Usb = 3,
}

impl BootDevice {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::HardDisk),
            1 => Some(Self::Cdrom),
            2 => Some(Self::Network),
            3 => Some(Self::Usb),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [BootDevice; 4] = [
        Self::HardDisk, Self::Cdrom, Self::Network, Self::Usb,
    ];
}

impl fmt::Display for BootDevice {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// Tests
// ===========================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn vm_state_roundtrip() {
        for v in VmState::ALL {
            let tag = v.to_tag();
            let decoded = VmState::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(VmState::from_tag(8).is_none());
    }

    #[test]
    fn virt_operation_roundtrip() {
        for v in VirtOperation::ALL {
            let tag = v.to_tag();
            let decoded = VirtOperation::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(VirtOperation::from_tag(11).is_none());
    }

    #[test]
    fn disk_format_roundtrip() {
        for v in DiskFormat::ALL {
            let tag = v.to_tag();
            let decoded = DiskFormat::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(DiskFormat::from_tag(5).is_none());
    }

    #[test]
    fn network_type_roundtrip() {
        for v in NetworkType::ALL {
            let tag = v.to_tag();
            let decoded = NetworkType::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(NetworkType::from_tag(4).is_none());
    }

    #[test]
    fn boot_device_roundtrip() {
        for v in BootDevice::ALL {
            let tag = v.to_tag();
            let decoded = BootDevice::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(BootDevice::from_tag(4).is_none());
    }

}

File: protocols/proven-virt/ffi/zig/src/virt.zig
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// virt.zig -- Zig FFI implementation of proven-virt.
//
// Implements the virtualization management state machine with:
//   - 64-slot mutex-protected VM session pool
//   - VM lifecycle state machine with transition validation
//   - Disk, network, and boot device configuration
//   - Resource tracking (vCPUs, memory)
//   - Operation validation against current VM state
//   - Thread-safe via per-pool mutex
//
// All exported functions use C calling convention (callconv(.c)) and
// communicate state via u8 tags matching VirtABI.Types exactly.

const std = @import("std");

// =========================================================================
// Enums (matching VirtABI.Types tag assignments)
// =========================================================================

/// VM lifecycle states (ABI tags 0-7).
pub const VMState = enum(u8) {
    creating = 0,
    running = 1,
    paused = 2,
    suspended = 3,
    shutting_down = 4,
    stopped = 5,
    crashed = 6,
    migrating = 7,
};

/// VM operations (ABI tags 0-10).
pub const Operation = enum(u8) {
    create = 0,
    start = 1,
    stop = 2,
    restart = 3,
    pause = 4,
    resume_ = 5,
    suspend_ = 6,
    migrate = 7,
    snapshot = 8,
    clone = 9,
    delete = 10,
};

/// Disk image formats (ABI tags 0-4).
pub const DiskFormat = enum(u8) {
    raw = 0,
    qcow2 = 1,
    vdi = 2,
    vmdk = 3,
    vhd = 4,
};

/// Virtual network types (ABI tags 0-3).
pub const NetworkType = enum(u8) {
    nat = 0,
    bridged = 1,
    internal = 2,
    host_only = 3,
};

/// Boot devices (ABI tags 0-3).
pub const BootDevice = enum(u8) {
    hard_disk = 0,
    cdrom = 1,
    network = 2,
    usb = 3,
};

// =========================================================================
// Internal data structures
// =========================================================================

/// Maximum concurrent VM sessions.
const MAX_SESSIONS: usize = 64;

/// Maximum VM name length.
const MAX_NAME_LEN: usize = 256;

/// Maximum migration destination length.
const MAX_DEST_LEN: usize = 512;

/// A virtual machine session.
const Session = struct {
    /// Current VM lifecycle state.
    state: VMState,
    /// VM name.
    name: [MAX_NAME_LEN]u8,
    name_len: u32,
    /// Number of virtual CPUs.
    vcpus: u16,
    /// Memory in megabytes.
    memory_mb: u32,
    /// Disk format.
    disk_format: DiskFormat,
    /// Network type.
    net_type: NetworkType,
    /// Boot device.
    boot_device: BootDevice,
    /// Migration destination (valid only during migration).
    migrate_dest: [MAX_DEST_LEN]u8,
    migrate_dest_len: u32,
    /// Number of snapshots taken.
    snapshot_count: u32,
    /// Whether this session slot is in use.
    active: bool,
};

/// Default (empty) session.
const empty_session: Session = .{
    .state = .stopped,
    .name = [_]u8{0} ** MAX_NAME_LEN,
    .name_len = 0,
    .vcpus = 1,
    .memory_mb = 512,
    .disk_format = .qcow2,
    .net_type = .nat,
    .boot_device = .hard_disk,
    .migrate_dest = [_]u8{0} ** MAX_DEST_LEN,
    .migrate_dest_len = 0,
    .snapshot_count = 0,
    .active = false,
};

// =========================================================================
// Global state
// =========================================================================

var sessions: [MAX_SESSIONS]Session = [_]Session{empty_session} ** MAX_SESSIONS;
var mutex: std.Thread.Mutex = .{};

// =========================================================================
// Internal helpers
// =========================================================================

/// Validate a slot index, returning null if out of range or inactive.
fn validSlot(slot: c_int) ?usize {
    if (slot < 0 or slot >= MAX_SESSIONS) return null;
    const idx: usize = @intCast(slot);
    if (!sessions[idx].active) return null;
    return idx;
}

// =========================================================================
// Exported C ABI functions
// =========================================================================

/// Returns the ABI version number.
pub export fn virt_abi_version() callconv(.c) u32 {
    return 1;
}

/// Create a new VM. Returns slot index (>=0) or -1 on failure.
pub export fn virt_create(
    name_ptr: [*]const u8,
    name_len: u32,
    vcpus: u16,
    memory_mb: u32,
    disk_fmt: u8,
    net_type: u8,
    boot_dev: u8,
) callconv(.c) c_int {
    mutex.lock();
    defer mutex.unlock();

    if (name_len == 0 or name_len > MAX_NAME_LEN) return -1;
    if (vcpus == 0) return -1;
    if (memory_mb == 0) return -1;
    if (disk_fmt > 4) return -1;
    if (net_type > 3) return -1;
    if (boot_dev > 3) return -1;

    for (&sessions, 0..) |*s, i| {
        if (!s.active) {
            s.* = empty_session;
            @memcpy(s.name[0..name_len], name_ptr[0..name_len]);
            s.name_len = name_len;
            s.vcpus = vcpus;
            s.memory_mb = memory_mb;
            s.disk_format = @enumFromInt(disk_fmt);
            s.net_type = @enumFromInt(net_type);
            s.boot_device = @enumFromInt(boot_dev);
            s.state = .creating;
            s.active = true;
            return @intCast(i);
        }
    }
    return -1;
}

/// Destroy a VM slot.
pub export fn virt_destroy(slot: c_int) callconv(.c) void {
    mutex.lock();
    defer mutex.unlock();
    if (slot < 0 or slot >= MAX_SESSIONS) return;
    sessions[@intCast(slot)] = empty_session;
}

/// Returns the current VMState tag for a VM.
pub export fn virt_state(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 5; // stopped fallback
    return @intFromEnum(sessions[idx].state);
}

/// Start a VM. Returns 0 on success, 1 on rejection.
/// Valid from Creating or Stopped.
pub export fn virt_start(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    const state = sessions[idx].state;
    if (state == .creating or state == .stopped) {
        sessions[idx].state = .running;
        return 0;
    }
    return 1;
}

/// Stop a VM. Returns 0 on success, 1 on rejection.
/// Valid from Running.
pub export fn virt_stop(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .running) return 1;
    sessions[idx].state = .stopped;
    return 0;
}

/// Pause a VM. Returns 0 on success, 1 on rejection.
/// Valid from Running.
pub export fn virt_pause(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .running) return 1;
    sessions[idx].state = .paused;
    return 0;
}

/// Resume a paused VM. Returns 0 on success, 1 on rejection.
/// Valid from Paused.
pub export fn virt_resume(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .paused) return 1;
    sessions[idx].state = .running;
    return 0;
}

/// Suspend a VM to disk. Returns 0 on success, 1 on rejection.
/// Valid from Running.
pub export fn virt_suspend(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .running) return 1;
    sessions[idx].state = .suspended;
    return 0;
}

/// Restart a running VM. Returns 0 on success, 1 on rejection.
/// Valid from Running (transitions Running -> Running).
pub export fn virt_restart(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .running) return 1;
    // Restart: remains running (simulated stop+start)
    return 0;
}

/// Begin live migration. Returns 0 on success, 1 on rejection.
/// Valid from Running.
pub export fn virt_migrate_begin(
    slot: c_int,
    dest_ptr: [*]const u8,
    dest_len: u32,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .running) return 1;
    if (dest_len == 0 or dest_len > MAX_DEST_LEN) return 1;

    @memcpy(sessions[idx].migrate_dest[0..dest_len], dest_ptr[0..dest_len]);
    sessions[idx].migrate_dest_len = dest_len;
    sessions[idx].state = .migrating;
    return 0;
}

/// Complete migration. Returns 0 on success, 1 on rejection.
/// Transitions Migrating -> Running.
pub export fn virt_migrate_complete(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .migrating) return 1;

    sessions[idx].migrate_dest_len = 0;
    sessions[idx].state = .running;
    return 0;
}

/// Delete a VM. Returns 0 on success, 1 on rejection.
/// Valid from Stopped or Crashed.
pub export fn virt_delete(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    const state = sessions[idx].state;
    if (state == .stopped or state == .crashed) {
        sessions[idx] = empty_session;
        return 0;
    }
    return 1;
}

/// Check if an operation is valid from the given VM state.
/// Returns 1 if valid, 0 if not.
pub export fn virt_can_transition(from: u8, op: u8) callconv(.c) u8 {
    // from=VMState tag, op=Operation tag
    if (op == 1) { // Start
        if (from == 0 or from == 5) return 1; // Creating/Stopped
    }
    if (op == 2) { // Stop
        if (from == 1) return 1; // Running
    }
    if (op == 3) { // Restart
        if (from == 1) return 1; // Running
    }
    if (op == 4) { // Pause
        if (from == 1) return 1; // Running
    }
    if (op == 5) { // Resume
        if (from == 2) return 1; // Paused
    }
    if (op == 6) { // Suspend
        if (from == 1) return 1; // Running
    }
    if (op == 7) { // Migrate
        if (from == 1) return 1; // Running
    }
    if (op == 8) { // Snapshot
        if (from == 1 or from == 2 or from == 5) return 1; // Running/Paused/Stopped
    }
    if (op == 9) { // Clone
        if (from == 5) return 1; // Stopped
    }
    if (op == 10) { // Delete
        if (from == 5 or from == 6) return 1; // Stopped/Crashed
    }
    return 0;
}

/// Returns vCPU count for a VM.
pub export fn virt_vcpu_count(slot: c_int) callconv(.c) u16 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return sessions[idx].vcpus;
}

/// Returns memory (MB) for a VM.
pub export fn virt_memory_mb(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return sessions[idx].memory_mb;
}

/// Returns disk format tag for a VM.
pub export fn virt_disk_format(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return @intFromEnum(sessions[idx].disk_format);
}

/// Returns number of active VM sessions.
pub export fn virt_session_count() callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();
    var count: u32 = 0;
    for (&sessions) |*s| {
        if (s.active) count += 1;
    }
    return count;
}
